### ISSUE TITLE
ifs_same_cond bails out on equal methods that return bool 

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#10272]
### Summary

apparently we detect  boolan expressions of the form `x == y`  but we don't detect `fn(&self) -> bool` ones???
I get it that this can be iffy if sideeffects are involved but in this example the fn does not even need `&mut self` :/

Could we somehow warn here  for const-fns at least maybe?

### Lint Name

ifs_same_cond

### Reproducer

I tried this code:

```rust
pub fn main() {
    let a = Some(32);
    if let Some(32) = a {
        let b = String::from("x");
        if b.contains("ah") { // same as
            eprintln!("AAA");
        } else if b.contains("ah") { // here
            println!("AAA");
        }
    }
}

```

I expected to see this happen:
clippy warning

Instead, this happened:

### Version

```text
rustc 1.69.0-nightly (dc1d9d50f 2023-01-31)
binary: rustc
commit-hash: dc1d9d50fba2f6a1ccab8748a0050cde38253f60
commit-date: 2023-01-31
host: x86_64-unknown-linux-gnu
release: 1.69.0-nightly
LLVM version: 15.0.7
```

**Comment by J-ZhengLi** [comment]
claim first, question it later~ @rustbot claim

**Comment by J-ZhengLi** [comment]
just want to confirm one thing before I start: what will be the side-effect even if the caller wasn't declared mutable?
Seems like there's already a [test case here](https://github.com/rust-lang/rust-clippy/blob/5b6795f50b521b11393d39f823d8a3e5b262b134/tests/ui/ifs_same_cond.rs#L40-L43) in the file, but for some reason the person though it will be ok to leave it as is, so I'm not sure...

**Comment by J-ZhengLi** [comment]
well, here comes the question 🤣:
If there's a mut pointer, and most of its methods are `const fn`, should the same method calls trigger this lint? like:

```rust
let p: *mut i8 = std::ptr::null_mut();
if p.is_null(){
    unsafe { p.write(..) }
}
else if p.is_null{..}
else {..}
```
